R/lint.R

```diff
@@ -42,38 +42,24 @@
 lint <- function(filename, linters = NULL, ..., cache = FALSE, parse_settings = TRUE, text = NULL) {
   # TODO(next release after 3.0.0): remove this deprecated workaround
   dots <- list(...)
-  if (length(dots) > 0L && is.logical(dots[[1L]]) && !nzchar(names2(dots)[1L])) {
+  if (has_positional_logical(dots)) {
     warning(
       "'cache' is no longer available as a positional argument; please supply 'cache' as a named argument instead. ",
       "This warning will be upgraded to an error in the next release."
     )
     cache <- dots[[1L]]
     dots <- dots[-1L]
   }
-  if (is.null(text)) {
-    inline_data <- rex::re_matches(filename, rex::rex(newline))
-    if (inline_data) {
-      text <- gsub("\n$", "", filename)
-      lines <- strsplit(text, "\n", fixed = TRUE)[[1L]]
-      filename <- NULL
-    } else {
-      lines <- read_lines(filename)
-    }
-  } else {
-    inline_data <- TRUE
-    if (length(text) > 1L) {
-      text <- paste(text, collapse = "\n")
-    }
-    lines <- strsplit(text, "\n", fixed = TRUE)[[1L]]
-  }
 
-  no_filename <- missing(filename) || is.null(filename)
+  needs_tempfile <- missing(filename) || rex::re_matches(filename, rex::rex(newline))
+  inline_data <- !is.null(text) || needs_tempfile
+  lines <- get_lines(filename, text)
 
-  if (inline_data && no_filename) {
+  if (needs_tempfile) {
```

Comment:
do this in a helper? helper returns temp filename so `on.exit(unlink())` happens fluidly

Reply:
How do you mean it?

Reply:
```r
# helper
write_text_to_tempfile <- function(lines) {
  filename <- tempfile()
  con <- file(filename, open = "w", encoding = settings$encoding)
  on.exit(close(con))
  writeLines(text = lines, con = con, sep = "\n")
  filename
}

# usage
if (needs_tempfile) {
  filename <- write_text_to_tempfile(lines)
  on.exit(unlink(lines), add = TRUE
}
```

i guess that doesn't help cyclocomp unless we consume the `if()` into the helper :|

Reply:
Yeah, unfortunately it doesn't.
I mean we could be mean and do an `on.exit()` on `parent.frame()`, but I think that's a code smell not worth pursuing.